### PR TITLE
Run validations on `save`

### PR DESCRIPTION
**Pull request by rubenrails**
In an effort to replicate the way active model validations work, this will run validations on regular `save`, returning `false` on validation error. There's also an option to skip validations altogether by using `save(validate: false)`.

@seancookr @javierg Review needed.

### CODE REVIEW

File: lib/dolly/document.rb [hunk 1]
```diff
@@ -51,7 +51,8 @@ def rev= value
       doc['_rev'] = value
     end
 
-    def save
+    def save options = {}
+      return false unless options[:validate] == false || valid?
```
Comment (seancookr) [inline lib/dolly/document.rb]: Id like to understand the usecase for the `options` a bit before I ok this

Comment (netlogics) [inline lib/dolly/document.rb]: @seancookr this is consistent with activerecords implementation. Consistent behavior seems like a good use case and brings dolly a step closer to implementing validations. https://github.com/rails/rails/blob/master/activerecord/lib/active_record/validations.rb#L36

@rubenrails please include touch as an option as well. Crude example
```
if timestamps
  set_created_at
  set_updated_at unless options[:touch] == false
```

Comment (javierg) [inline lib/dolly/document.rb]: We are not implementing validations, that is not dolly concern, webare just giving the app an interface for implementing it. Dolly should work independently of rails

Comment (seancookr) [inline lib/dolly/document.rb]: :point_up: what @javierg said.

Comment (netlogics) [inline lib/dolly/document.rb]: Definitely, dolly should work independently of rails but implementing validations and rails independence aren't mutually exclusive imo. Regardless, I wasn't trying to suggest implementing validations rather offering a use case for making our interfaces and methods such as `save` consistent with what a rails developer would expect.

Comment (rubenrails) [inline lib/dolly/document.rb]: > what @javierg said.
I don't get the point. Following a convention doesn't mean we're not working "independently" from rails IMO.

TL;DR What @netlogics  said.

This PR only standardizes the return value of the `save` method to a boolean. In case you have some sort of validation in place (which responds to `valid?`) you'll be fine. The option to bypass validations was exactly to give a wider range of devs the option to save without validating in a per basis case.

Comment (seancookr) [inline lib/dolly/document.rb]: @rubenrails that was my response to @netlogics regarding the validation stuff.  I hadn't responded to you yet since you had not answered yet.  I agree with both you and @netlogics reasonings to included the options.  I was waiting to respond to you after hearing you reasons, as they may have actually been different than @netlogics response

Comment (javierg) [inline lib/dolly/document.rb]: @rubenrails @netlogics Agree on conventions, just trying to keep the focus. Because "implemting" have a meaning... On this particular PR I don't see the harm

**Comment by rubenrails** [comment]
@seancookr Do @netlogics' explanation and mine answer your question?

> this is consistent with activerecords implementation. Consistent behavior seems like a good use case

>  to give a wider range of devs the option to save without validating in a per basis case.

Anything else?

**Comment by seancookr** [comment]
> I agree with both you and @netlogics reasonings to included the options.

:shipit:
